src/soc/sophgo/cv180x.c

```diff
@@ -0,0 +1,396 @@
+/*
+  Copyright (c) 2024 Shenzhen Milk-V Technology Co., Ltd
+  Author: Willian <willian@milkv.io>
+          Carbon <carbon@milkv.io>
+
+  This Source Code Form is subject to the terms of the Mozilla Public
+  License, v. 2.0. If a copy of the MPL was not distributed with this
+  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+*/
+
+#include <sys/mman.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+
+#include "cv180x.h"
+#include "../../wiringx.h"
+#include "../soc.h"
+#include "../../i2c-dev.h"
+
+#define CV180X_GPIO_GROUP_COUNT 4
+
+const static uintptr_t gpio_register_physical_address[MAX_REG_AREA] = {0x03020000, 0x03021000, 0x03022000, 0x05021000};
+#define GPIO_SWPORTA_DR		0x000	
+#define GPIO_SWPORTA_DDR		0x004
+#define GPIO_EXT_PORTA		0x050
+
+static uintptr_t pinmux_register_virtual_address = NULL;
+
+#define PINMUX_BASE		0x03001000	// pinmux group 1
+
+#define CLEAR_BITS(addr, bit, size) \
+	(*addr = *addr & ~(~(-1 << size) << bit) | (~(-1 << size) << bit << REGISTER_WRITE_MASK))
+#define GET_BITS(addr, bit, size) \
+	((*addr & ~(-1 << size) << bit) >> (bit - size))
+
+struct soc_t *cv180x = NULL;
+
+static struct layout_t {
+	char *name;
+	int gpio_group;
+	int num;
+
+	struct {
+		unsigned long offset;
+		unsigned long value;
+	} pinmux;
+
+	struct {
+		unsigned long offset;
+		unsigned long bit;
+	} direction;
+
+	struct {
+		unsigned long offset;
+		unsigned long bit;
+	} data;
+
+	int support;
+	enum pinmode_t mode;
+	int fd;
+} layout[] = {
+	{"XGPIOA_7", 0, 487, {0x0, 0x3}, {GPIO_SWPORTA_DDR, 7}, {GPIO_SWPORTA_DR, 7}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_8", 0, 488, {0x4, 0x3}, {GPIO_SWPORTA_DDR, 8}, {GPIO_SWPORTA_DR, 8}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_9", 0, 489, {0x8, 0x3}, {GPIO_SWPORTA_DDR, 9}, {GPIO_SWPORTA_DR, 9}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_10", 0, 490, {0xc, 0x3}, {GPIO_SWPORTA_DDR, 10}, {GPIO_SWPORTA_DR, 10}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_11", 0, 491, {0x10, 0x3}, {GPIO_SWPORTA_DDR, 11}, {GPIO_SWPORTA_DR, 11}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_12", 0, 492, {0x14, 0x3}, {GPIO_SWPORTA_DDR, 12}, {GPIO_SWPORTA_DR, 12}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_13", 0, 493, {0x18, 0x3}, {GPIO_SWPORTA_DDR, 13}, {GPIO_SWPORTA_DR, 13}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"XGPIOA_14", 0, 494, {0x1c, 0x3}, {GPIO_SWPORTA_DDR, 14}, {GPIO_SWPORTA_DR, 14}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0}, /* GP14 */
```

Comment:
```suggestion
	{"XGPIOA_14", 0, 494, {0x1c, 0x3}, {GPIO_SWPORTA_DDR, 14}, {GPIO_SWPORTA_DR, 14}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
```
你这个注释跟cv180x SoC毫无关系，不要放进来